runtime/expr/eval.go

```diff
@@ -766,8 +766,20 @@ func indexMap(zctx *zed.Context, ectx Context, typ *zed.TypeMap, mapBytes zcode.
 		return zctx.Missing()
 	}
 	if key.Type != typ.KeyType {
-		// XXX issue #3360
-		return zctx.Missing()
+		if union, ok := zed.TypeUnder(typ.KeyType).(*zed.TypeUnion); ok {
+			for _, t := range union.Types {
+				if key.Type == t {
```

Comment:
union.TagOf does roughly the same thing we're doing here, so we can simplify a little by calling it and checking for a nonnegative result.
```go
			if tag := union.TagOf(key.Type); tag >= 0 {
```